File: nodejs/src/cloud/cloudSession.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import type { AssistantMessageEvent } from "../session.js";
import type {
    CloudAskUserResponsePayload,
    CloudElicitationResponsePayload,
    CloudModeSwitchPayload,
    CloudPermissionResponsePayload,
    CloudPlanApprovalResponsePayload,
    CloudSessionEvent,
    CloudSessionEventHandler,
    CloudSessionEventPayload,
    CloudSessionEventType,
    CloudSessionMetadata,
    ElicitationResult,
    ExitPlanModeResult,
    MessageOptions,
    MissionControlCommandType,
    TypedCloudSessionEventHandler,
} from "../types.js";
import { MissionControlCommandType as CommandType } from "../types.js";
import type { MissionControlClient } from "./missionControlClient.js";

const DEFAULT_POLL_INTERVAL_MS = 5_000;
const DEFAULT_INITIAL_EVENT_TIMEOUT_MS = 10_000;
const DEFAULT_INITIAL_EVENT_POLL_INTERVAL_MS = 500;

export interface CloudSessionCreateOptions {
    client: MissionControlClient;
    metadata: CloudSessionMetadata;
    pollIntervalMs?: number;
    initialEventTimeoutMs?: number;
    initialEventPollIntervalMs?: number;
    onEventPollError?: (error: Error) => void;
}

export class CloudSession {
    private readonly client: MissionControlClient;
    private readonly pollIntervalMs: number;
    private readonly initialEventTimeoutMs: number;
    private readonly initialEventPollIntervalMs: number;
    private readonly onEventPollError?: (error: Error) => void;
    private readonly eventHandlers = new Set<CloudSessionEventHandler>();
    private readonly typedEventHandlers = new Map<
        CloudSessionEventType,
        Set<(event: CloudSessionEvent) => void>
    >();
    private readonly events: CloudSessionEvent[] = [];
    private seenEventIds = new Set<string>();
    private seenEventIdsAtLastTimestamp = new Set<string>();
    private lastSeenTimestamp?: string;
    private eventPoller: ReturnType<typeof setInterval> | undefined;
    private isPolling = false;
    private isDisconnected = false;
    private remoteSteerable = true;

    readonly sessionId: string;
    readonly metadata: CloudSessionMetadata;

    constructor(options: CloudSessionCreateOptions) {
        this.client = options.client;
        this.metadata = options.metadata;
        this.sessionId = options.metadata.missionControlSessionId ?? options.metadata.taskId;
        this.pollIntervalMs = options.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;
        this.initialEventTimeoutMs =
            options.initialEventTimeoutMs ?? DEFAULT_INITIAL_EVENT_TIMEOUT_MS;
        this.initialEventPollIntervalMs =
            options.initialEventPollIntervalMs ?? DEFAULT_INITIAL_EVENT_POLL_INTERVAL_MS;
        this.onEventPollError = options.onEventPollError;
    }

    async connect(): Promise<void> {
        const initialEvents = await this.waitForInitialEvents();
        this.recordEvents(initialEvents);
        this.startEventPolling();
    }

    on<K extends CloudSessionEventType>(
        eventType: K,
        handler: TypedCloudSessionEventHandler<K>
    ): () => void;
    on(handler: CloudSessionEventHandler): () => void;
    on<K extends CloudSessionEventType>(
        eventTypeOrHandler: K | CloudSessionEventHandler,
        handler?: TypedCloudSessionEventHandler<K>
    ): () => void {
        if (typeof eventTypeOrHandler === "string" && handler) {
            const eventType = eventTypeOrHandler;
            if (!this.typedEventHandlers.has(eventType)) {
                this.typedEventHandlers.set(eventType, new Set());
            }
            const storedHandler = handler as (event: CloudSessionEvent) => void;
            this.typedEventHandlers.get(eventType)!.add(storedHandler);
            return () => {
                this.typedEventHandlers.get(eventType)?.delete(storedHandler);
            };
        }

        const wildcardHandler = eventTypeOrHandler as CloudSessionEventHandler;
        this.eventHandlers.add(wildcardHandler);
        return () => {
            this.eventHandlers.delete(wildcardHandler);
        };
    }

    async send(options: MessageOptions): Promise<void> {
        this.assertConnected();
        await this.submitRemoteCommand(CommandType.UserMessage, options.prompt);
    }

    async sendAndWait(
        options: MessageOptions,
        timeout?: number
    ): Promise<AssistantMessageEvent | undefined> {
        const effectiveTimeout = timeout ?? 60_000;
        let lastAssistantMessage: AssistantMessageEvent | undefined;
        let timeoutId: ReturnType<typeof setTimeout> | undefined;
        let unsubscribe: (() => void) | undefined;

        const idlePromise = new Promise<void>((resolve, reject) => {
            unsubscribe = this.on((event) => {
                if (event.type === "assistant.message") {
                    lastAssistantMessage = event as AssistantMessageEvent;
                } else if (event.type === "session.idle") {
                    resolve();
                } else if (event.type === "session.error") {
                    reject(new Error(event.data.message));
                }
            });
        });

        try {
            await this.send(options);
            await Promise.race([
                idlePromise,
                new Promise<never>((_, reject) => {
                    timeoutId = setTimeout(
                        () =>
                            reject(
                                new Error(
                                    `Timeout after ${effectiveTimeout}ms waiting for session.idle`
                                )
                            ),
                        effectiveTimeout
                    );
                }),
            ]);
            return lastAssistantMessage;
        } finally {
            if (timeoutId !== undefined) {
                clearTimeout(timeoutId);
            }
            unsubscribe?.();
        }
    }

    async abort(): Promise<void> {
        this.assertConnected();
        await this.submitRemoteCommand(CommandType.Abort);
    }

    async submitRemoteCommand(type: MissionControlCommandType, content?: string): Promise<void> {
        this.assertConnected();
        if (!this.remoteSteerable) {
            throw new Error("This session is read-only — remote steering is not enabled");
        }
        await this.client.steerTask(this.metadata.taskId, { type, content });
    }

    async respondToPermission(payload: CloudPermissionResponsePayload): Promise<void> {
        await this.submitRemoteCommand(CommandType.PermissionResponse, JSON.stringify(payload));
    }

    async respondToAskUser(payload: CloudAskUserResponsePayload): Promise<void> {
        await this.submitRemoteCommand(CommandType.AskUserResponse, JSON.stringify(payload));
    }

    async respondToElicitation(payload: CloudElicitationResponsePayload): Promise<void> {
        await this.submitRemoteCommand(CommandType.ElicitationResponse, JSON.stringify(payload));
    }

    async respondToExitPlanMode(payload: CloudPlanApprovalResponsePayload): Promise<void> {
        await this.submitRemoteCommand(CommandType.PlanApprovalResponse, JSON.stringify(payload));
    }

    async switchMode(payload: CloudModeSwitchPayload): Promise<void> {
        await this.submitRemoteCommand(CommandType.ModeSwitch, JSON.stringify(payload));
    }

    async respondToElicitationResult(promptId: string, result: ElicitationResult): Promise<void> {
        await this.respondToElicitation({ promptId, ...result });
    }

    async respondToPlanApproval(promptId: string, result: ExitPlanModeResult): Promise<void> {
        await this.respondToExitPlanMode({ promptId, ...result });
    }

    getMessages(): CloudSessionEvent[] {
        return [...this.events];
    }

    async disconnect(): Promise<void> {
        this.stopEventPolling();
        this.eventHandlers.clear();
        this.typedEventHandlers.clear();
        this.isDisconnected = true;
    }

    async destroy(): Promise<void> {
        return this.disconnect();
    }

    async [Symbol.asyncDispose](): Promise<void> {
        return this.disconnect();
    }

    startEventPolling(): void {
        if (this.eventPoller || this.isDisconnected) {
            return;
        }

        this.eventPoller = setInterval(() => {
            this.pollEvents().catch((error) => this.reportPollError(error));
        }, this.pollIntervalMs);
        this.eventPoller.unref?.();
    }

    stopEventPolling(): void {
        if (this.eventPoller) {
            clearInterval(this.eventPoller);
            this.eventPoller = undefined;
        }
    }

    private async waitForInitialEvents(): Promise<CloudSessionEvent[]> {
        const deadline = Date.now() + this.initialEventTimeoutMs;
        while (true) {
            const events = await this.client.listTaskEvents(this.metadata.taskId);
            if (events.length > 0) {
                return CloudSession.sortEventsChronologically(events);
            }
            if (this.initialEventTimeoutMs <= 0 || Date.now() >= deadline) {
                return [];
            }
            await sleep(this.initialEventPollIntervalMs);
        }
    }

    private async pollEvents(): Promise<void> {
        if (this.isPolling || this.isDisconnected) {
            return;
        }

        this.isPolling = true;
        try {
            const events = await this.client.listTaskEvents(this.metadata.taskId);
            const newEvents = this.collectNewEvents(events);
            this.recordEvents(newEvents);
        } finally {
            this.isPolling = false;
        }
    }

    private collectNewEvents(events: CloudSessionEvent[]): CloudSessionEvent[] {
        const newEvents = events.filter((event) => {
            if (this.seenEventIds.has(event.id)) return false;
            if (!this.lastSeenTimestamp) return true;
            const order = event.timestamp.localeCompare(this.lastSeenTimestamp);
            if (order > 0) return true;
            if (order < 0) return false;
            return !this.seenEventIdsAtLastTimestamp.has(event.id);
        });

        return CloudSession.sortEventsChronologically(newEvents);
    }

    private recordEvents(events: CloudSessionEvent[]): void {
        for (const event of CloudSession.sortEventsChronologically(events)) {
            if (this.seenEventIds.has(event.id)) continue;
            this.seenEventIds.add(event.id);
            this.events.push(event);
            this.markEventAsSeenAtTimestamp(event);
            this.updateRemoteSteerable(event);
            this.dispatchEvent(event);
        }
    }

    private markEventAsSeenAtTimestamp(event: CloudSessionEvent): void {
        if (this.lastSeenTimestamp !== event.timestamp) {
            this.lastSeenTimestamp = event.timestamp;
            this.seenEventIdsAtLastTimestamp = new Set();
        }
        this.seenEventIdsAtLastTimestamp.add(event.id);
    }

    private updateRemoteSteerable(event: CloudSessionEvent): void {
        if (event.type === "session.remote_steerable_changed") {
            this.remoteSteerable = event.data.remoteSteerable;
        }
    }

    private dispatchEvent(event: CloudSessionEvent): void {
        const typedHandlers = this.typedEventHandlers.get(event.type);
        if (typedHandlers) {
            for (const handler of typedHandlers) {
                try {
                    handler(event as CloudSessionEventPayload<typeof event.type>);
                } catch {
                    // Keep one failing handler from stopping event polling.
                }
            }
        }

        for (const handler of this.eventHandlers) {
            try {
                handler(event);
            } catch {
                // Keep one failing handler from stopping event polling.
            }
        }
    }

    private reportPollError(error: unknown): void {
        const normalized = error instanceof Error ? error : new Error(String(error));
        if (this.onEventPollError) {
            this.onEventPollError(normalized);
        }
    }

    private assertConnected(): void {
        if (this.isDisconnected) {
            throw new Error("Cloud session is disconnected");
        }
    }

    private static sortEventsChronologically(events: CloudSessionEvent[]): CloudSessionEvent[] {
        return [...events].sort(
            (left, right) =>
                left.timestamp.localeCompare(right.timestamp) || left.id.localeCompare(right.id)
        );
    }
}

function sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

File: python/copilot/cloud/cloud_session.py
"""
Cloud session — remote-control client for a sandbox-backed cloud session.

The :class:`CloudSession` class polls Mission Control for task events and
exposes methods for sending user messages and steering commands.
"""

from __future__ import annotations

import asyncio
import json
import logging
from collections.abc import Callable
from typing import Any, overload

from .mission_control_client import MissionControlClient
from .types import (
    CloudAskUserResponsePayload,
    CloudElicitationResponsePayload,
    CloudModeSwitchPayload,
    CloudPermissionResponsePayload,
    CloudPlanApprovalResponsePayload,
    CloudSessionEvent,
    CloudSessionEventHandler,
    CloudSessionMetadata,
    ElicitationResult,
    ExitPlanModeResult,
    MissionControlCommandType,
)

logger = logging.getLogger(__name__)

_DEFAULT_POLL_INTERVAL_S = 5.0
_DEFAULT_INITIAL_EVENT_TIMEOUT_S = 10.0
_DEFAULT_INITIAL_EVENT_POLL_INTERVAL_S = 0.5


class CloudSession:
    """Remote-control client for a cloud sandbox session.

    After construction, call :meth:`connect` to start receiving events.

    Args:
        client: The Mission Control HTTP client.
        metadata: Metadata describing the cloud session/task.
        poll_interval_ms: Milliseconds between event polls (default: 5000).
        initial_event_timeout_ms: Milliseconds to wait for the first event (default: 10000).
        initial_event_poll_interval_ms: Milliseconds between polls while waiting for the
            first event (default: 500).
        on_event_poll_error: Callback invoked when an event poll fails.
    """

    def __init__(
        self,
        *,
        client: MissionControlClient,
        metadata: CloudSessionMetadata,
        poll_interval_ms: int | None = None,
        initial_event_timeout_ms: int | None = None,
        initial_event_poll_interval_ms: int | None = None,
        on_event_poll_error: Callable[[Exception], None] | None = None,
    ) -> None:
        self._client = client
        self.metadata = metadata
        self.session_id = metadata.mission_control_session_id or metadata.task_id

        self._poll_interval_s = (
            poll_interval_ms / 1000 if poll_interval_ms is not None else _DEFAULT_POLL_INTERVAL_S
        )
        self._initial_event_timeout_s = (
            initial_event_timeout_ms / 1000
            if initial_event_timeout_ms is not None
            else _DEFAULT_INITIAL_EVENT_TIMEOUT_S
        )
        self._initial_event_poll_interval_s = (
            initial_event_poll_interval_ms / 1000
            if initial_event_poll_interval_ms is not None
            else _DEFAULT_INITIAL_EVENT_POLL_INTERVAL_S
        )
        self._on_event_poll_error = on_event_poll_error

        self._event_handlers: set[CloudSessionEventHandler] = set()
        self._typed_event_handlers: dict[str, set[CloudSessionEventHandler]] = {}
        self._events: list[CloudSessionEvent] = []
        self._seen_event_ids: set[str] = set()
        self._seen_event_ids_at_last_timestamp: set[str] = set()
        self._last_seen_timestamp: str | None = None
        self._poller_task: asyncio.Task[None] | None = None
        self._is_polling = False
        self._is_disconnected = False
        self._remote_steerable = True

    # ------------------------------------------------------------------
    # Connection lifecycle
    # ------------------------------------------------------------------

    async def connect(self) -> None:
        """Connect to the cloud session and start event polling."""
        initial_events = await self._wait_for_initial_events()
        self._record_events(initial_events)
        self._start_event_polling()

    # ------------------------------------------------------------------
    # Event subscription
    # ------------------------------------------------------------------

    @overload
    def on(self, handler: CloudSessionEventHandler, /) -> Callable[[], None]: ...

    @overload
    def on(self, event_type: str, handler: CloudSessionEventHandler, /) -> Callable[[], None]: ...

    def on(
        self,
        event_type_or_handler: str | CloudSessionEventHandler,
        handler: CloudSessionEventHandler | None = None,
        /,
    ) -> Callable[[], None]:
        """Register an event handler.

        Can be called with a wildcard handler::

            session.on(lambda event: print(event.type))

        Or with a specific event type::

            session.on("session.idle", lambda event: print("idle"))

        Returns a callable that removes the handler when invoked.
        """
        if isinstance(event_type_or_handler, str) and handler is not None:
            event_type = event_type_or_handler
            if event_type not in self._typed_event_handlers:
                self._typed_event_handlers[event_type] = set()
            self._typed_event_handlers[event_type].add(handler)

            def _unsubscribe() -> None:
                handlers = self._typed_event_handlers.get(event_type)
                if handlers:
                    handlers.discard(handler)

            return _unsubscribe

        wildcard_handler = event_type_or_handler
        assert callable(wildcard_handler)
        self._event_handlers.add(wildcard_handler)

        def _unsubscribe() -> None:
            self._event_handlers.discard(wildcard_handler)

        return _unsubscribe

    # ------------------------------------------------------------------
    # Sending messages & steering
    # ------------------------------------------------------------------

    async def send(self, *, prompt: str) -> None:
        """Send a user message to the cloud session.

        Args:
            prompt: The message text to send.
        """
        self._assert_connected()
        await self.submit_remote_command(MissionControlCommandType.USER_MESSAGE, prompt)

    async def send_and_wait(
        self,
        *,
        prompt: str,
        timeout: float | None = None,
    ) -> CloudSessionEvent | None:
        """Send a message and wait for the session to reach idle.

        Returns the last ``assistant.message`` event received before idle,
        or ``None`` if the session became idle without an assistant message.

        Args:
            prompt: The message text to send.
            timeout: Maximum seconds to wait (default: 60).
        """
        effective_timeout = timeout if timeout is not None else 60.0
        last_assistant_message: CloudSessionEvent | None = None
        done = asyncio.Event()
        error_holder: list[Exception] = []

        def _handler(event: CloudSessionEvent) -> None:
            nonlocal last_assistant_message
            if event.type == "assistant.message":
                last_assistant_message = event
            elif event.type == "session.idle":
                done.set()
            elif event.type == "session.error":
                msg = event.data.get("message", "Unknown error") if event.data else "Unknown error"
                error_holder.append(Exception(msg))
                done.set()

        unsubscribe = self.on(_handler)
        try:
            await self.send(prompt=prompt)
            await asyncio.wait_for(done.wait(), timeout=effective_timeout)
            if error_holder:
                raise error_holder[0]
            return last_assistant_message
        except TimeoutError as exc:
            raise TimeoutError(
                f"Timeout after {effective_timeout}s waiting for session.idle"
            ) from exc
        finally:
            unsubscribe()

    async def abort(self) -> None:
        """Abort the current cloud session operation."""
        self._assert_connected()
        await self.submit_remote_command(MissionControlCommandType.ABORT)

    async def submit_remote_command(
        self,
        command_type: MissionControlCommandType,
        content: str | None = None,
    ) -> None:
        """Send a raw steering command to Mission Control.

        Args:
            command_type: The type of steering command.
            content: Optional payload content.
        """
        self._assert_connected()
        if not self._remote_steerable:
            raise RuntimeError("This session is read-only — remote steering is not enabled")
        request: dict[str, Any] = {"type": command_type.value}
        if content is not None:
            request["content"] = content
        await self._client.steer_task(self.metadata.task_id, request)

    # ------------------------------------------------------------------
    # Response helpers
    # ------------------------------------------------------------------

    async def respond_to_permission(self, payload: CloudPermissionResponsePayload) -> None:
        """Respond to a permission request."""
        wire = _to_camel_case_dict(payload)
        await self.submit_remote_command(
            MissionControlCommandType.PERMISSION_RESPONSE, json.dumps(wire)
        )

    async def respond_to_ask_user(self, payload: CloudAskUserResponsePayload) -> None:
        """Respond to an ask-user prompt."""
        wire = _to_camel_case_dict(payload)
        await self.submit_remote_command(
            MissionControlCommandType.ASK_USER_RESPONSE, json.dumps(wire)
        )

    async def respond_to_elicitation(self, payload: CloudElicitationResponsePayload) -> None:
        """Respond to an elicitation prompt."""
        wire = _to_camel_case_dict(payload)
        await self.submit_remote_command(
            MissionControlCommandType.ELICITATION_RESPONSE, json.dumps(wire)
        )

    async def respond_to_exit_plan_mode(self, payload: CloudPlanApprovalResponsePayload) -> None:
        """Respond to a plan approval prompt."""
        wire = _to_camel_case_dict(payload)
        await self.submit_remote_command(
            MissionControlCommandType.PLAN_APPROVAL_RESPONSE, json.dumps(wire)
        )

    async def switch_mode(self, payload: CloudModeSwitchPayload) -> None:
        """Switch the cloud session mode."""
        await self.submit_remote_command(
            MissionControlCommandType.MODE_SWITCH, json.dumps(dict(payload))
        )

    async def respond_to_elicitation_result(
        self, prompt_id: str, result: ElicitationResult
    ) -> None:
        """Convenience: respond to an elicitation with a prompt ID and result."""
        payload = CloudElicitationResponsePayload(
            prompt_id=prompt_id,
            **result,  # type: ignore[typeddict-item]
        )
        await self.respond_to_elicitation(payload)

    async def respond_to_plan_approval(self, prompt_id: str, result: ExitPlanModeResult) -> None:
        """Convenience: respond to a plan approval with a prompt ID and result."""
        payload = CloudPlanApprovalResponsePayload(
            prompt_id=prompt_id,
            **result,  # type: ignore[typeddict-item]
        )
        await self.respond_to_exit_plan_mode(payload)

    # ------------------------------------------------------------------
    # Event access
    # ------------------------------------------------------------------

    def get_messages(self) -> list[CloudSessionEvent]:
        """Return a copy of all events received so far."""
        return list(self._events)

    # ------------------------------------------------------------------
    # Disconnect
    # ------------------------------------------------------------------

    async def disconnect(self) -> None:
        """Disconnect from the cloud session and stop event polling."""
        self._stop_event_polling()
        self._event_handlers.clear()
        self._typed_event_handlers.clear()
        self._is_disconnected = True

    async def destroy(self) -> None:
        """Alias for :meth:`disconnect`."""
        await self.disconnect()

    async def __aenter__(self) -> CloudSession:
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: Any,
    ) -> None:
        await self.disconnect()

    # ------------------------------------------------------------------
    # Event polling internals
    # ------------------------------------------------------------------

    def _start_event_polling(self) -> None:
        if self._poller_task is not None or self._is_disconnected:
            return

        self._poller_task = asyncio.ensure_future(self._poll_loop())

    def _stop_event_polling(self) -> None:
        if self._poller_task is not None:
            self._poller_task.cancel()
            self._poller_task = None

    async def _poll_loop(self) -> None:
        try:
            while not self._is_disconnected:
                await asyncio.sleep(self._poll_interval_s)
                try:
                    await self._poll_events()
                except Exception as exc:
                    self._report_poll_error(exc)
        except asyncio.CancelledError:
            pass

    async def _wait_for_initial_events(self) -> list[CloudSessionEvent]:
        deadline = asyncio.get_event_loop().time() + self._initial_event_timeout_s
        while True:
            events = await self._client.list_task_events(self.metadata.task_id)
            if events:
                return _sort_events_chronologically(events)
            if self._initial_event_timeout_s <= 0 or asyncio.get_event_loop().time() >= deadline:
                return []
            await asyncio.sleep(self._initial_event_poll_interval_s)

    async def _poll_events(self) -> None:
        if self._is_polling or self._is_disconnected:
            return
        self._is_polling = True
        try:
            events = await self._client.list_task_events(self.metadata.task_id)
            new_events = self._collect_new_events(events)
            self._record_events(new_events)
        finally:
            self._is_polling = False

    def _collect_new_events(self, events: list[CloudSessionEvent]) -> list[CloudSessionEvent]:
        new: list[CloudSessionEvent] = []
        for event in events:
            if event.id in self._seen_event_ids:
                continue
            if self._last_seen_timestamp is None:
                new.append(event)
                continue
            order = _compare_strings(event.timestamp, self._last_seen_timestamp)
            if order > 0:
                new.append(event)
            elif order == 0 and event.id not in self._seen_event_ids_at_last_timestamp:
                new.append(event)
        return _sort_events_chronologically(new)

    def _record_events(self, events: list[CloudSessionEvent]) -> None:
        for event in _sort_events_chronologically(events):
            if event.id in self._seen_event_ids:
                continue
            self._seen_event_ids.add(event.id)
            self._events.append(event)
            self._mark_event_as_seen_at_timestamp(event)
            self._update_remote_steerable(event)
            self._dispatch_event(event)

    def _mark_event_as_seen_at_timestamp(self, event: CloudSessionEvent) -> None:
        if self._last_seen_timestamp != event.timestamp:
            self._last_seen_timestamp = event.timestamp
            self._seen_event_ids_at_last_timestamp = set()
        self._seen_event_ids_at_last_timestamp.add(event.id)

    def _update_remote_steerable(self, event: CloudSessionEvent) -> None:
        if event.type == "session.remote_steerable_changed" and event.data:
            self._remote_steerable = bool(event.data.get("remoteSteerable", True))

    def _dispatch_event(self, event: CloudSessionEvent) -> None:
        typed_handlers = self._typed_event_handlers.get(event.type)
        if typed_handlers:
            for handler in list(typed_handlers):
                try:
                    handler(event)
                except Exception:
                    pass

        for handler in list(self._event_handlers):
            try:
                handler(event)
            except Exception:
                pass

    def _report_poll_error(self, error: Exception) -> None:
        if self._on_event_poll_error:
            self._on_event_poll_error(error)

    def _assert_connected(self) -> None:
        if self._is_disconnected:
            raise RuntimeError("Cloud session is disconnected")


# ------------------------------------------------------------------
# Module-level helpers
# ------------------------------------------------------------------


def _sort_events_chronologically(
    events: list[CloudSessionEvent],
) -> list[CloudSessionEvent]:
    return sorted(events, key=lambda e: (e.timestamp, e.id))


def _compare_strings(a: str, b: str) -> int:
    if a > b:
        return 1
    if a < b:
        return -1
    return 0


def _snake_to_camel(name: str) -> str:
    """Convert a snake_case name to camelCase."""
    parts = name.split("_")
    return parts[0] + "".join(p.capitalize() for p in parts[1:])


def _to_camel_case_dict(d: dict[str, Any]) -> dict[str, Any]:
    """Convert a dict with snake_case keys to camelCase keys."""
    return {_snake_to_camel(k): v for k, v in d.items()}
